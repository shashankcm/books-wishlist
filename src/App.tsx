import React from "react";
import "./styles/App.scss";
import BookSearch from "./book-search/BookSearch";
import BookWishList from "./book-wishlist/BookWishList";

function App() {
  return (
    <main>
      <BookSearch />
      <BookWishList />
    </main>
  );
}

export default App;
