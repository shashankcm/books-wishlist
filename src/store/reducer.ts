import {
  Wish,
  WishlistAction,
  ActionTypes,
  AllBooks,
  AvailableBooksAction,
} from "../shared/types";

export const wishlistReducer = (state: Wish[], action: WishlistAction) => {
  switch (action.type) {
    case ActionTypes.ADD_TO_WISHLIST:
      return [
        ...state,
        {
          id: action.payload.id,
          title: action.payload.title,
        },
      ];

    default:
      return state;
  }
};

export const availableBooksReducer = (
  state: AllBooks,
  action: AvailableBooksAction
) => {
  switch (action.type) {
    case ActionTypes.ADD_AVAILABLE_BOOKS:
      return Object.assign(
        {},
        {
          kind: action.payload.kind,
          totalItems: action.payload.totalItems,
          items: action.payload.items,
        }
      );

    default:
      return state;
  }
};
