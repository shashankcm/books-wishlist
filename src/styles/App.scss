@import "./variables";
@import "./Book.scss";

* {
  box-sizing: border-box;
}

body {
  font-family: "Architects Daughter", cursive;
}

h1,
h2 {
  font-size: 24px;
  margin: $spacing-xxs 0;
  font-weight: 300;
}

form {
  margin-bottom: $spacing-xs;
}

input:focus {
  outline: none;
}

input {
  border: none;
}

main {
  padding: $spacing-xs $spacing-xxs;
  display: flex;
  height: 100vh;

  @include md {
    flex-direction: column;
  }
}

.book__search {
  flex: 0.8;
  margin: 20px;

  &--search__form {
    width: 100%;
    border: 2px solid black;
    padding: 20px;
  }

  &--search__input {
    flex: 1;
    width: 100%;
  }

  &--container {
    display: grid;
    width: 100%;
    overflow-y: scroll;
    grid-template-columns: 1fr;
    grid-template-rows: 1fr;
    border: 2px solid black;
    padding: 20px;
    min-height: 83%;

    &::-webkit-scrollbar {
      display: none;
    }

    @include lg {
      min-height: 83%;
    }

    @include sm {
      padding: 5px;
    }
  }
}

.book__wishlist {
  flex: 0.2;
  min-width: 300px;
  margin: 0 20px;

  @include md {
    margin: 60px 20px;
  }

  @include md {
    margin: 60px 20px;
  }

  &--reading__list {
    border: 2px solid black;
    text-align: center;
    -webkit-box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
    -moz-box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
    box-shadow: 10px 10px 5px 0px rgba(0, 0, 0, 0.75);
  }

  &--selected__book {
    width: 100%;
    height: 90%;
    border: 2px solid black;
    padding: 20px;
  }
}
