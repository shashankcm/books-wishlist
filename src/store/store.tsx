import React, { createContext, useContext, useReducer } from "react";
import { IState, IContextProps, StoreProviderProps } from "../shared/types";
import { wishlistReducer, availableBooksReducer } from "./reducer";

export const initialState: IState = {
  wishlist: [],
  availableBooks: { kind: "", totalItems: 0, items: [] },
};

export const StoreContext = createContext<IContextProps>({
  state: initialState,
  dispatch: () => null,
});

const reducer = (state: IState, action: any) => ({
  wishlist: wishlistReducer(state.wishlist, action),
  availableBooks: availableBooksReducer(state.availableBooks, action),
});

export const StoreProvider = ({ children }: StoreProviderProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <StoreContext.Provider value={{ state, dispatch }}>
      {children}
    </StoreContext.Provider>
  );
};

export const useStoreProviderValue = () => useContext(StoreContext);
