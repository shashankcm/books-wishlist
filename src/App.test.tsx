import React from "react";
import { render, cleanup } from "@testing-library/react";
import App from "./App";
import { StoreProvider } from "./store/store";

export const renderWithContext = (component: React.ReactNode) => {
  return {
    ...render(<StoreProvider>{component}</StoreProvider>),
  };
};

const renderApplication = () =>
  render(
    <StoreProvider>
      <App />
    </StoreProvider>
  );

afterEach(cleanup);

describe("App Component", () => {
  it("should take a snapshot", () => {
    const { asFragment } = renderApplication();
    expect(
      asFragment(
        <StoreProvider>
          <App />
        </StoreProvider>
      )
    ).toMatchSnapshot();
  });
});
