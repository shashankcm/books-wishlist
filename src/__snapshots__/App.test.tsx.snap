// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should take a snapshot 1`] = `
<DocumentFragment>
  <main>
    <div
      class="book__search"
    >
      <form
        class="book__search--search__form"
      >
        <fieldset>
          <legend>
            Search Books
          </legend>
          <label
            for="book_search"
          >
            Book Name
          </label>
          <input
            aria-label="bookSearch"
            class="book__search--search__input"
            id="book_search"
            name="book search"
            placeholder="Search for books to add to your reading list and press Enter"
            type="search"
            value=""
          />
        </fieldset>
      </form>
      <div
        class="book__search--container"
      />
    </div>
    <div
      class="book__wishlist"
    >
      <h3
        class="book__wishlist--reading__list"
      >
        My Reading Wishlist (
        <span
          data-testid="numberOfWishlistBooks"
        >
          0
        </span>
        )
      </h3>
      <div
        class="book__wishlist--selected__book"
        data-testid="listOfBooks"
      >
        <ol />
      </div>
    </div>
  </main>
</DocumentFragment>
`;
